torch/csrc/autograd/python_function.cpp

```diff
@@ -1051,6 +1100,5 @@ std::shared_ptr<PyFunction> THPFunction_asFunction(THPFunction* self)
     return std::shared_ptr<PyFunction>();
   }
 
-  Py_INCREF((PyObject*)self);
-  return std::shared_ptr<PyFunction>(&self->cdata, Decref());
+  return self->cdata.lock();
```

Comment:
How come we no longer need to do the incref/decref here? IIRC debug builds of Python report refcounting errors if we don't keep the C++/Python refcounts approximately in sync.

Reply:
I don't know. I don't understand why it was previously needed. I guess I can build a debug build of Python and see if I still get refcounting errors. If I don't, would that be sufficient?

Reply:
I have a debug build of PyTorch with this patch and it has not reported any refcounting errors.

Reply:
Not a debug build of PyTorch, a debug build _of CPython_. I once made it such that we always use [a single `shared_ptr` to manage `PyFunction`](https://github.com/pytorch/pytorch/pull/1454), which has [started triggering Python's GC asserts](https://github.com/pytorch/pytorch/issues/1626) and the refcounting scheme [has been updated by @colesbury](https://github.com/pytorch/pytorch/pull/1705) accordingly. We should check that this patch does not suffer from the same issues as mine.

Also, there used to be this warning that you're not supposed to take weak pointers of `PyFunction`s because they can be managed by multiple shared_ptr blocks. That's no longer the case after this patch, right?